Add SandiaPC Pressure Fitter

nprisbrey opened the pull request:
Applies the SandiaPC pressure fitter and makes the 'rho0_guess' input mandatory when using said fitter (as requested by Christine). Addresses Issue #7.

nprisbrey committed:
Started Filling in Functions for SandiaPC

nprisbrey committed:
Implemented The SandiaPC Fitter

nprisbrey committed:
Made 'rho0_guess' Mandatory When Using Fitter 'SandiaPC'

leek2 commented:
Nathan, can you add a test in examples that shows this fitter working?  

I think that Christine wanted rho0 to not move during fitting (not just requiring that a guess be provided.  A guess can move during fitting.)  Can your test also be sure to verify that?  (You might need 2 tests).  

In general we need more tests.  It would be nice to have unit testing and stuff, but for now examples will do.

nprisbrey commented:
Hey Jim, thanks for your comment. I agree on the comment of rho0 moving. I'll look into changing that . If rho0 is not to move, should I make a new command line argument that receives rho0 and doesn't move?

I agree on your stance of unit testing, though it does look like we're going to use examples. I assume that you are referencing 1 dimensional examples. Where do you suggest that I find/create the data?

leek2 commented:
I'd rather not add a new command line argument.  If you can figure out how to take rho0 as a guess, but not move it as part of the fitting that seems best.  We would just add a note about that in the documentation of the Sandia Pc fit.  (As a comment and in the README. That's another part of this, you have to put in the documentation before you can say you're done.)

Well, how well does the Sandia Pc fit the existing example1D data?  If that works use it.  If not ask Carrie if she has any.  (She might know of a paper that uses this.) If she doesn't have anything that works, synthesize some data by just making your own inputs to the Sandia Pc curve, getting points from it, and make sure that the fit correctly matches your original coefficients.    

You could even perturb the input data a bit so that the fit has to work a bit, but should come up with the same fit.

nprisbrey committed:
Switched 'k0' To 'kneg0' In SandiaPC To Avoid Confusion With Bulk Modulus

nprisbrey committed:
Made 'rho0' a Fixed Parameter in SandiaPC Fitter

nprisbrey committed:
Finished Documentation Changes

nprisbrey commented:
Everything should be wrapped up -- documentation, commenting, rho0_guess being required while using the SandiaPC fitter, and rho0 not changing while SandiaPC is being optimized. What are your thoughts, @leek2? Are we ready to move forward?

nprisbrey commented on curvallis/curve_editing/curve_fitters.py:
```diff
@@ -378,11 +388,11 @@ def guess_coefficients(self, points):
-            if (hasattr (self, 'k0')):
+            if (not hasattr (self, 'k0')):
                 self.k0 = 1.0e11
-            if (hasattr (self, 'k0_prime')):
+            if (not hasattr (self, 'k0_prime')):
                 self.k0_prime = 20.0
-            if (hasattr (self, 'rho0')):
+            if (not hasattr (self, 'rho0')):
```
These three lines of code seem to have been written incorrectly according to the comments. Fixing this code brings it into accordance with the comments and allows the SandiaPC fitter to function appropriately without an external part of code changing the user-given rho0 value.

leek2 commented:
1. Where's the example/test file for sanidapc?  I couldn't find it.
2. Also, it looks like rho0 moved from my guess:

[leek2@localhost examples]$ ../curvallis.py --input_file=1dexample1.dat --fit_type=sandiapc --rho0_guess=1.0

Current directory is "/home/leek2/Curvallis/examples"
Reading data from "1dexample1.dat"
Lines read: 154
---------------------------------------------------------------
input points: 1dexample1.dat:
Number of points: 153
X range: 1.000000000000000E-10 .. 1.000000000000000E+04
Y range: -2.333203892760288E-01 .. 1.101335000000000E+07
---------------------------------------------------------------
Creating region 0 for X range 1.000000000000000E-10 .. 1.000000000000000E+04
REGION "0" INFO:
    X Range: 1.000000000000000E-10 .. 1.000000000000000E+04
    Line sets count: 1
        Line set "1dexample1.dat":            Name: "1dexample1.dat"
            Point Count: 153

Approximate Fit (No ghost points yet):
kneg1 = -2.16532541649e-05;
k0 = 0.454661612648;
k1 = -57.6432662973;
k2 = 984.732241279;
k3 = 7354.06598953;
k4 = -536324.375423;
k5 = 5221325.05305;
rho0 = 5.0;
regions:
kneg1 = -0.000230955012683;
k0 = 0.454952223949;
k1 = -5.41044449971;
k2 = 8.66949030421;
k3 = 6.07179581864;
k4 = -41.5467045704;
k5 = 37.946152055;
rho0 = 5.0;

nprisbrey committed:
Replaced Code That Was Overwrote

It seems that the "not"s in the code were overwritten by the last commit, which commit was purposed to only change documentation.

nprisbrey commented:
Thanks for the comment @leek2 . You are correct in your observation. It looks like a bit of the code was overwritten in commit 5c99828. If the rho0 continues to change now then I will need to look farther into the problem, though the code online should now be the same that was working on my computer earlier this evening.

As for the example/test file, I believe that the two existing 1D data examples that are already present in the examples directory are sufficient examples and the SandiaPC fitter fits the data well, notwithstanding whatever rho0 value you'd like to give. Would you still like me to fabricate more?

leek2 commented:
It's true that the example data fits fine.  What I'm thinking of is an example .ini file.  It should give the fit_type as SandiaPc, and a good rho0 guess.  It can use 1dexample1.dat as data.  Then a comment explaining things.  
It serves as both a simple test, and an example.  (I would delete all the extraneous, commented out options that you see in 1dexample1.ini)

nprisbrey committed:
SandiaPC Fitter Example .ini Files Added

nprisbrey commented:
I just added two ".ini" files: one for each 1D example ".dat" file in the examples directory. I modelled them and tested them after the previous ".ini" files that were already present. I'm not sure what you mean by a comment explaining things. If wanted, could you give some more detail on this @leek2?
